# Maximizing Engagement: Insights on Hacker News Post Timing and Topic Impact

## Introduction

In this data science project, we aim to analyze the engagement patterns of posts on the popular technology site Hacker News. Specifically, we will compare two distinct types of posts: "Ask HN" and "Show HN." "Ask HN" posts are questions posed to the Hacker News community, while "Show HN" posts are used to showcase projects or interesting content. Our primary goal is to determine which type of post garners more comments on average.

To achieve this goal, we will analyze a dataset containing 20,000 submissions to Hacker News. We will filter the dataset to focus only on posts with titles that begin with either "Ask HN" or "Show HN." By examining the number of comments each post receives, we will calculate and compare the average comment counts for both types of posts. Additionally, we will investigate whether the time of post creation influences the number of comments a post receives, providing insights into optimal posting times.

Our analysis will reveal whether "Ask HN" or "Show HN" posts generate more community engagement in terms of comments. Furthermore, we will identify specific time frames that correlate with higher comment counts. These findings will offer valuable guidance for users looking to maximize the visibility and interaction of their posts on Hacker News.

In [1]:
from csv import reader

opened_file = open('hacker_news.csv') #Open the dataset
read_file = reader(opened_file)
hn = list(read_file)                  #Convert the dataset into a list of lists
dataset_backup = hn                   #Create a data backup
headers = hn[0]                       #Extract the header row into a separate list
hn = hn[1:]                           #Omit the first row since it contains the columns' headers

print('The Hacker News dataset contains ', len(hn), ' rows.', '\n')
print('Headers row:', '\n', headers, '\n')
print('The first five rows from the dataset:', '\n',hn[:5])

The Hacker News dataset contains  20100  rows. 

Headers row: 
 ['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'] 

The first five rows from the dataset: 
 [['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nyt

### Dataset Overview

To provide a clear understanding of the dataset used in our analysis, here are the basic details:
- Total Number of Rows: 20,100
- Column Headers:
    - id: Unique identifier for each post
    - title: Title of the post
    - url: URL the post links to (if applicable)
    - num_points: Total points the post acquired (upvotes minus downvotes)
    - num_comments: Number of comments on the post
    - author: Username of the person who submitted the post
    - created_at: Date and time of the post's submission

This dataset provides a comprehensive view of Hacker News submissions, allowing us to analyze various aspects such as the number of comments, points received, and the timing of posts for different types of submissions ("Ask HN," "Show HN," and others).

In [2]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
    
print ('Number of "Ask HN" posts: ', len(ask_posts))
print ('Number of "Show HN" posts: ', len(show_posts))
print ('Number of other posts: ', len(other_posts))

Number of "Ask HN" posts:  1744
Number of "Show HN" posts:  1162
Number of other posts:  17194


In [3]:
total_ask_comments = 0
total_show_comments = 0

for row in ask_posts:
    comments = int(row[4])
    total_ask_comments += comments
avg_ask_comments = total_ask_comments / len(ask_posts)

for row in show_posts:
    comments = int(row[4])
    total_show_comments += comments
avg_show_comments = total_show_comments / len(show_posts)

print('The total numbers of comments on the "Ask HN" post is: ', total_ask_comments, '\n')
print ('The average number of comments on "Ask HN" posts is: ', round(avg_ask_comments, 2), '\n')

print('The total numbers of comments on the "Show HN" post is: ', total_show_comments, '\n')
print ('The average number of comments on "Show HN" posts is: ', round(avg_show_comments, 2), '\n')

The total numbers of comments on the "Ask HN" post is:  24483 

The average number of comments on "Ask HN" posts is:  14.04 

The total numbers of comments on the "Show HN" post is:  11988 

The average number of comments on "Show HN" posts is:  10.32 



In our analysis of the Hacker News dataset, we examined two distinct types of posts: "Ask HN" and "Show HN." Here are our key findings:
1. Our analysis showed that there are more "Ask HN" posts (1,744) than "Show HN" posts (1,162) on Hacker News.
2. "Ask HN" posts, where users ask questions to the community, tend to be more engaging, receiving a total of 24,483 comments and an average of 14 comments per post.
3. In comparison, "Show HN" posts, where users share projects or interesting content, received a total of 11,988 comments, averaging about 10 comments per post.
4. This suggests that asking questions on Hacker News tends to generate more community interaction than showcasing projects.

In [4]:
import datetime as dt

result_list = []

for row in ask_posts:
    date_and_comments_list = []
    date = row[6]
    comments = int(row[4])
    date_and_comments_list.append(date)
    date_and_comments_list.append(comments)
    result_list.append(date_and_comments_list)
print(result_list[:5])

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3], ['10/15/2015 16:38', 17]]


In [5]:
counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    date = row[0]
    comments = row[1]
    date_dt = dt.datetime.strptime(date, "%m/%d/%Y %H:%M")
    hour_str = date_dt.strftime("%H")
    
    if hour_str not in counts_by_hour:
        counts_by_hour[hour_str] = 1
        comments_by_hour[hour_str] = comments
    else:
        counts_by_hour[hour_str] += 1
        comments_by_hour[hour_str] += comments

print(counts_by_hour, '\n')
print(comments_by_hour, '\n')

{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58} 

{'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641} 



In [6]:
avg_by_hour = []

for row in comments_by_hour:
    hour = row
    comments = int(comments_by_hour[row])
    posts_count = counts_by_hour[row]
    average_comments = round((comments / posts_count), 2)
    avg_by_hour.append([hour, average_comments])
    
print(avg_by_hour)

[['09', 5.58], ['13', 14.74], ['10', 13.44], ['14', 13.23], ['16', 16.8], ['23', 7.99], ['12', 9.41], ['17', 11.46], ['15', 38.59], ['21', 16.01], ['20', 21.52], ['02', 23.81], ['18', 13.2], ['03', 7.8], ['05', 10.09], ['19', 10.8], ['01', 11.38], ['22', 6.75], ['08', 10.25], ['04', 7.17], ['00', 8.13], ['06', 9.02], ['07', 7.85], ['11', 11.05]]


In [7]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
print(swap_avg_by_hour)

[[5.58, '09'], [14.74, '13'], [13.44, '10'], [13.23, '14'], [16.8, '16'], [7.99, '23'], [9.41, '12'], [11.46, '17'], [38.59, '15'], [16.01, '21'], [21.52, '20'], [23.81, '02'], [13.2, '18'], [7.8, '03'], [10.09, '05'], [10.8, '19'], [11.38, '01'], [6.75, '22'], [10.25, '08'], [7.17, '04'], [8.13, '00'], [9.02, '06'], [7.85, '07'], [11.05, '11']]


In [8]:
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print(sorted_swap)

[[38.59, '15'], [23.81, '02'], [21.52, '20'], [16.8, '16'], [16.01, '21'], [14.74, '13'], [13.44, '10'], [13.23, '14'], [13.2, '18'], [11.46, '17'], [11.38, '01'], [11.05, '11'], [10.8, '19'], [10.25, '08'], [10.09, '05'], [9.41, '12'], [9.02, '06'], [8.13, '00'], [7.99, '23'], [7.85, '07'], [7.8, '03'], [7.17, '04'], [6.75, '22'], [5.58, '09']]


In [9]:
print("Top 5 Hours for Ask Posts Comments: ", '\n')

for row in sorted_swap[:5]:
    average_comments = row[0]
    time = row[1]
    time_dt = dt.datetime.strptime(time, "%H")
    time_str = time_dt.strftime("%H:%M")

    print(time_str, ': ', average_comments, ' average comments per post.')

Top 5 Hours for Ask Posts Comments:  

15:00 :  38.59  average comments per post.
02:00 :  23.81  average comments per post.
20:00 :  21.52  average comments per post.
16:00 :  16.8  average comments per post.
21:00 :  16.01  average comments per post.


In our detailed analysis of the Hacker News dataset, we investigated the optimal posting times for "Ask HN" posts to determine when they receive the most comments. Here are our key findings:

Our analysis revealed the best times to post questions ("Ask HN" posts) on Hacker News for maximum comments. The top five times are:
- 3 PM: 38.6 average comments
- 2 AM: 23.8 average comments
- 8 PM: 21.5 average comments
- 4 PM: 16.8 average comments
- 9 PM: 16 average comments

Posting questions at these times, especially at 3 PM, significantly increases the likelihood of receiving more comments and engagement from the community.

In [10]:
total_ask_points = 0
total_show_points = 0

for row in ask_posts:
    post_points = int(row[3])
    total_ask_points += post_points
    
for row in show_posts:
    post_points = int(row[3])
    total_show_points += post_points

avg_ask_points = total_ask_points / len(ask_posts)
avg_show_points = total_show_points / len(show_posts)

print ('Number of "Ask HN" posts: ', len(ask_posts))
print('The total numbers of points on the "Ask HN" post is: ', total_ask_points)
print ('The average number of points on "Ask HN" posts is: ', round(avg_ask_points, 2), '\n')

print ('Number of "Show HN" posts: ', len(show_posts))
print('The total numbers of points on the "Show HN" post is: ', total_show_points)
print ('The average number of points on "Show HN" posts is: ', round(avg_show_points, 2), '\n')

Number of "Ask HN" posts:  1744
The total numbers of points on the "Ask HN" post is:  26268
The average number of points on "Ask HN" posts is:  15.06 

Number of "Show HN" posts:  1162
The total numbers of points on the "Show HN" post is:  32019
The average number of points on "Show HN" posts is:  27.56 



In our extended analysis of the Hacker News dataset, we examined not only the comments but also the points (upvotes minus downvotes) received by "Ask HN" and "Show HN" posts. Here are our key findings:

- Our analysis revealed differences in how "Ask HN" and "Show HN" posts are received by the Hacker News community in terms of points (upvotes minus downvotes).
- "Ask HN" posts:
    - There are 1,744 of these posts.
    - They received a total of 26,268 points, averaging about 15 points per post.
- "Show HN" posts:
    - There are 1,162 of these posts.
    - They received a total of 32,019 points, averaging about 28 points per post.
- This suggests that while "Ask HN" posts are popular for generating comments, "Show HN" posts generally receive more upvotes, indicating greater community appreciation for shared projects or interesting content.

In [11]:
ask_posts_date_points = []
show_posts_date_points = []

for row in ask_posts:
    date_and_points_list = []
    date = row[6]
    points = int(row[3])
    date_and_points_list.append(date)
    date_and_points_list.append(points)
    ask_posts_date_points.append(date_and_points_list)

for row in show_posts:
    date_and_points_list = []
    date = row[6]
    points = int(row[3])
    date_and_points_list.append(date)
    date_and_points_list.append(points)
    show_posts_date_points.append(date_and_points_list)
    
print(ask_posts_date_points[:5], '\n')
print(show_posts_date_points[:5])

[['8/16/2016 9:55', 2], ['11/22/2015 13:43', 28], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 1], ['10/15/2015 16:38', 28]] 

[['11/25/2015 14:03', 26], ['11/29/2015 22:46', 747], ['4/28/2016 18:05', 1], ['7/28/2016 7:11', 3], ['1/9/2016 20:45', 1]]


In [12]:
ask_posts_counts_by_hour = {}
ask_posts_points_by_hour = {}

for row in ask_posts_date_points:
    date = row[0]
    points = row[1]
    date_dt = dt.datetime.strptime(date, "%m/%d/%Y %H:%M")
    hour_str = date_dt.strftime("%H")
    
    if hour_str not in ask_posts_counts_by_hour:
        ask_posts_counts_by_hour[hour_str] = 1
        ask_posts_points_by_hour[hour_str] = points
    else:
        ask_posts_counts_by_hour[hour_str] += 1
        ask_posts_points_by_hour[hour_str] += points

print(ask_posts_counts_by_hour, '\n')
print(ask_posts_points_by_hour, '\n')

show_posts_counts_by_hour = {}
show_posts_points_by_hour = {}

for row in show_posts_date_points:
    date = row[0]
    points = row[1]
    date_dt = dt.datetime.strptime(date, "%m/%d/%Y %H:%M")
    hour_str = date_dt.strftime("%H")
    
    if hour_str not in show_posts_counts_by_hour:
        show_posts_counts_by_hour[hour_str] = 1
        show_posts_points_by_hour[hour_str] = points
    else:
        show_posts_counts_by_hour[hour_str] += 1
        show_posts_points_by_hour[hour_str] += points

print(show_posts_counts_by_hour, '\n')
print(show_posts_points_by_hour)

{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58} 

{'09': 329, '13': 2062, '10': 1102, '14': 1282, '16': 2522, '23': 581, '12': 782, '17': 1941, '15': 3479, '21': 1721, '20': 1151, '02': 793, '18': 1741, '03': 374, '05': 552, '19': 1513, '01': 700, '22': 511, '08': 515, '04': 389, '00': 451, '06': 591, '07': 361, '11': 825} 

{'14': 86, '22': 46, '18': 61, '07': 26, '20': 60, '05': 19, '16': 93, '19': 55, '15': 78, '03': 27, '17': 93, '06': 16, '02': 30, '13': 99, '08': 34, '21': 47, '04': 26, '11': 44, '12': 61, '23': 36, '09': 30, '01': 28, '10': 36, '00': 31} 

{'14': 2187, '22': 1856, '18': 2215, '07': 494, '20': 1819, '05': 104, '16': 2634, '19': 1702, '15': 2228, '03': 679, '17': 2521, '06': 375, '02': 340, '13': 2438, '08': 519, '21': 866, '04': 386, '11': 1480, '12': 2543, '23': 152

In [13]:
ask_posts_avg_points_by_hour = []

for row in ask_posts_points_by_hour:
    hour = row
    points = int(ask_posts_points_by_hour[row])
    ask_posts_count = ask_posts_counts_by_hour[row]
    average_points = round((points / ask_posts_count), 2)
    ask_posts_avg_points_by_hour.append([hour, average_points])
    
print(ask_posts_avg_points_by_hour, '\n')

show_posts_avg_points_by_hour = []

for row in show_posts_points_by_hour:
    hour = row
    points = int(show_posts_points_by_hour[row])
    show_posts_count = show_posts_counts_by_hour[row]
    average_points = round((points / show_posts_count), 2)
    show_posts_avg_points_by_hour.append([hour, average_points])
    
print(show_posts_avg_points_by_hour)

[['09', 7.31], ['13', 24.26], ['10', 18.68], ['14', 11.98], ['16', 23.35], ['23', 8.54], ['12', 10.71], ['17', 19.41], ['15', 29.99], ['21', 15.79], ['20', 14.39], ['02', 13.67], ['18', 15.97], ['03', 6.93], ['05', 12.0], ['19', 13.75], ['01', 11.67], ['22', 7.2], ['08', 10.73], ['04', 8.28], ['00', 8.2], ['06', 13.43], ['07', 10.62], ['11', 14.22]] 

[['14', 25.43], ['22', 40.35], ['18', 36.31], ['07', 19.0], ['20', 30.32], ['05', 5.47], ['16', 28.32], ['19', 30.95], ['15', 28.56], ['03', 25.15], ['17', 27.11], ['06', 23.44], ['02', 11.33], ['13', 24.63], ['08', 15.26], ['21', 18.43], ['04', 14.85], ['11', 33.64], ['12', 41.69], ['23', 42.39], ['09', 18.43], ['01', 25.0], ['10', 18.92], ['00', 37.84]]


In [14]:
swap_ask_posts_avg_points_by_hour = []

for row in ask_posts_avg_points_by_hour:
    swap_ask_posts_avg_points_by_hour.append([row[1], row[0]])
    
print(swap_ask_posts_avg_points_by_hour, '\n')

swap_show_posts_avg_points_by_hour = []

for row in show_posts_avg_points_by_hour:
    swap_show_posts_avg_points_by_hour.append([row[1], row[0]])
    
print(swap_show_posts_avg_points_by_hour, '\n')

[[7.31, '09'], [24.26, '13'], [18.68, '10'], [11.98, '14'], [23.35, '16'], [8.54, '23'], [10.71, '12'], [19.41, '17'], [29.99, '15'], [15.79, '21'], [14.39, '20'], [13.67, '02'], [15.97, '18'], [6.93, '03'], [12.0, '05'], [13.75, '19'], [11.67, '01'], [7.2, '22'], [10.73, '08'], [8.28, '04'], [8.2, '00'], [13.43, '06'], [10.62, '07'], [14.22, '11']] 

[[25.43, '14'], [40.35, '22'], [36.31, '18'], [19.0, '07'], [30.32, '20'], [5.47, '05'], [28.32, '16'], [30.95, '19'], [28.56, '15'], [25.15, '03'], [27.11, '17'], [23.44, '06'], [11.33, '02'], [24.63, '13'], [15.26, '08'], [18.43, '21'], [14.85, '04'], [33.64, '11'], [41.69, '12'], [42.39, '23'], [18.43, '09'], [25.0, '01'], [18.92, '10'], [37.84, '00']] 



In [15]:
sorted_swap_ask = sorted(swap_ask_posts_avg_points_by_hour, reverse = True)
print(sorted_swap_ask, '\n')

sorted_swap_show = sorted(swap_show_posts_avg_points_by_hour, reverse = True)
print(sorted_swap_show)

[[29.99, '15'], [24.26, '13'], [23.35, '16'], [19.41, '17'], [18.68, '10'], [15.97, '18'], [15.79, '21'], [14.39, '20'], [14.22, '11'], [13.75, '19'], [13.67, '02'], [13.43, '06'], [12.0, '05'], [11.98, '14'], [11.67, '01'], [10.73, '08'], [10.71, '12'], [10.62, '07'], [8.54, '23'], [8.28, '04'], [8.2, '00'], [7.31, '09'], [7.2, '22'], [6.93, '03']] 

[[42.39, '23'], [41.69, '12'], [40.35, '22'], [37.84, '00'], [36.31, '18'], [33.64, '11'], [30.95, '19'], [30.32, '20'], [28.56, '15'], [28.32, '16'], [27.11, '17'], [25.43, '14'], [25.15, '03'], [25.0, '01'], [24.63, '13'], [23.44, '06'], [19.0, '07'], [18.92, '10'], [18.43, '21'], [18.43, '09'], [15.26, '08'], [14.85, '04'], [11.33, '02'], [5.47, '05']]


In [16]:
print("Top 5 Hours for Ask Posts Points: ", '\n')

for row in sorted_swap_ask[:5]:
    average_points = row[0]
    time = row[1]
    time_dt = dt.datetime.strptime(time, "%H")
    time_str = time_dt.strftime("%H:%M")

    print(time_str, ': ', average_points, ' average points per post.')
    
print('\n')
print("Top 5 Hours for Show Posts Points: ", '\n')

for row in sorted_swap_show[:5]:
    average_points = row[0]
    time = row[1]
    time_dt = dt.datetime.strptime(time, "%H")
    time_str = time_dt.strftime("%H:%M")

    print(time_str, ': ', average_points, ' average points per post.')

Top 5 Hours for Ask Posts Points:  

15:00 :  29.99  average points per post.
13:00 :  24.26  average points per post.
16:00 :  23.35  average points per post.
17:00 :  19.41  average points per post.
10:00 :  18.68  average points per post.


Top 5 Hours for Show Posts Points:  

23:00 :  42.39  average points per post.
12:00 :  41.69  average points per post.
22:00 :  40.35  average points per post.
00:00 :  37.84  average points per post.
18:00 :  36.31  average points per post.


In our investigation of the Hacker News dataset, we analyzed the impact of post creation times on the number of points received by "Ask HN" and "Show HN" posts. Here are our key findings:

- Our analysis shows that the time of day when a post is made can influence its popularity in terms of points received (upvotes minus downvotes).

- Best times for "Ask HN" posts:
    - 3 PM: 30 points on average
    - 1 PM: 24 points on average
    - 4 PM: 23 points on average
    - 5 PM: 19 points on average
    - 10 AM: 19 points on average

- Best times for "Show HN" posts:
    - 11 PM: 42 points on average
    - 12 PM: 42 points on average
    - 10 PM: 40 points on average
    - Midnight: 38 points on average
    - 6 PM: 36 points on average

These results indicate that "Ask HN" posts receive the most points when submitted in the afternoon, particularly around 3 PM, while "Show HN" posts perform best when submitted late at night or around noon.

In [17]:
total_other_comments = 0

for row in other_posts:
    comments = int(row[4])
    total_other_comments += comments

avg_other_comments = total_other_comments / len(other_posts)

print('The total numbers of comments on other posts is: ', total_other_comments, '\n')
print ('The average number of comments on other posts is: ', round(avg_other_comments, 2), '\n')

The total numbers of comments on other posts is:  462055 

The average number of comments on other posts is:  26.87 



In [18]:
other_posts_date_and_comments = []

for row in other_posts:
    date_and_comments_list = []
    date = row[6]
    comments = int(row[4])
    date_and_comments_list.append(date)
    date_and_comments_list.append(comments)
    other_posts_date_and_comments.append(date_and_comments_list)

print(other_posts_date_and_comments[:5], '\n')

other_posts_counts_by_hour = {}
other_posts_comments_by_hour = {}

for row in other_posts_date_and_comments:
    date = row[0]
    comments = row[1]
    date_dt = dt.datetime.strptime(date, "%m/%d/%Y %H:%M")
    hour_str = date_dt.strftime("%H")
    
    if hour_str not in other_posts_counts_by_hour:
        other_posts_counts_by_hour[hour_str] = 1
        other_posts_comments_by_hour[hour_str] = comments
    else:
        other_posts_counts_by_hour[hour_str] += 1
        other_posts_comments_by_hour[hour_str] += comments

print(other_posts_counts_by_hour, '\n')
print(other_posts_comments_by_hour, '\n')

other_posts_comments_avg_by_hour = []

for row in other_posts_comments_by_hour:
    hour = row
    comments = int(other_posts_comments_by_hour[row])
    posts_count = other_posts_counts_by_hour[row]
    average_comments = round((comments / posts_count), 2)
    other_posts_comments_avg_by_hour.append([hour, average_comments])
    
print(other_posts_comments_avg_by_hour, '\n')

swap_other_posts_comments_avg_by_hour = []

for row in other_posts_comments_avg_by_hour:
    swap_other_posts_comments_avg_by_hour.append([row[1], row[0]])
print(swap_other_posts_comments_avg_by_hour, '\n')

sorted_swap_other_posts_comments_avg_by_hour = sorted(swap_other_posts_comments_avg_by_hour, reverse = True)
print(sorted_swap_other_posts_comments_avg_by_hour, '\n')

print("Top 5 Hours for Other Posts Comments: ", '\n')

for row in sorted_swap_other_posts_comments_avg_by_hour[:5]:
    average_comments = row[0]
    time = row[1]
    time_dt = dt.datetime.strptime(time, "%H")
    time_str = time_dt.strftime("%H:%M")

    print(time_str, ': ', average_comments, ' average comments per post.')

[['8/4/2016 11:52', 52], ['1/26/2016 19:30', 10], ['6/23/2016 22:20', 1], ['6/17/2016 0:01', 1], ['9/30/2015 4:12', 2]] 

{'11': 660, '19': 980, '22': 758, '00': 611, '04': 454, '09': 534, '16': 1101, '18': 1084, '10': 591, '12': 789, '20': 911, '03': 407, '17': 1169, '14': 958, '13': 918, '01': 500, '23': 674, '08': 496, '02': 441, '21': 874, '15': 1040, '06': 408, '05': 388, '07': 448} 

{'11': 19532, '19': 26167, '22': 17635, '00': 16544, '04': 10953, '09': 14732, '16': 27959, '18': 29186, '10': 15728, '12': 23944, '20': 21080, '03': 10918, '17': 32727, '14': 30973, '13': 28363, '01': 11536, '23': 16592, '08': 13405, '02': 12254, '21': 20635, '15': 30700, '06': 8714, '05': 9768, '07': 12010} 

[['11', 29.59], ['19', 26.7], ['22', 23.27], ['00', 27.08], ['04', 24.13], ['09', 27.59], ['16', 25.39], ['18', 26.92], ['10', 26.61], ['12', 30.35], ['20', 23.14], ['03', 26.83], ['17', 28.0], ['14', 32.33], ['13', 30.9], ['01', 23.07], ['23', 24.62], ['08', 27.03], ['02', 27.79], ['21', 23.6

In [19]:
total_other_points = 0

for row in other_posts:
    post_points = int(row[3])
    total_other_points += post_points
    
avg_other_points = total_other_points / len(other_posts)

print ('Number of other posts: ', len(other_posts))
print('The total numbers of points on other post is: ', total_other_points)
print ('The average number of points on other posts is: ', round(avg_other_points, 2), '\n')

other_posts_date_points = []

for row in other_posts:
    date_and_points_list = []
    date = row[6]
    points = int(row[3])
    date_and_points_list.append(date)
    date_and_points_list.append(points)
    other_posts_date_points.append(date_and_points_list)
    
print(other_posts_date_points[:5], '\n')

other_posts_counts_by_hour = {}
other_posts_points_by_hour = {}

for row in other_posts_date_points:
    date = row[0]
    points = row[1]
    date_dt = dt.datetime.strptime(date, "%m/%d/%Y %H:%M")
    hour_str = date_dt.strftime("%H")
    
    if hour_str not in other_posts_counts_by_hour:
        other_posts_counts_by_hour[hour_str] = 1
        other_posts_points_by_hour[hour_str] = points
    else:
        other_posts_counts_by_hour[hour_str] += 1
        other_posts_points_by_hour[hour_str] += points

print(other_posts_counts_by_hour, '\n')
print(other_posts_points_by_hour, '\n')

other_posts_avg_points_by_hour = []

for row in other_posts_points_by_hour:
    hour = row
    points = int(other_posts_points_by_hour[row])
    other_posts_count = other_posts_counts_by_hour[row]
    average_points = round((points / other_posts_count), 2)
    other_posts_avg_points_by_hour.append([hour, average_points])
    
print(other_posts_avg_points_by_hour, '\n')

swap_other_posts_avg_points_by_hour = []

for row in other_posts_avg_points_by_hour:
    swap_other_posts_avg_points_by_hour.append([row[1], row[0]])
    
print(swap_other_posts_avg_points_by_hour, '\n')

sorted_swap_other = sorted(swap_other_posts_avg_points_by_hour, reverse = True)
print(sorted_swap_other, '\n')

print("Top 5 Hours for Other Posts Points: ", '\n')

for row in sorted_swap_other[:5]:
    average_points = row[0]
    time = row[1]
    time_dt = dt.datetime.strptime(time, "%H")
    time_str = time_dt.strftime("%H:%M")

    print(time_str, ': ', average_points, ' average points per post.')

Number of other posts:  17194
The total numbers of points on other post is:  952664
The average number of points on other posts is:  55.41 

[['8/4/2016 11:52', 386], ['1/26/2016 19:30', 39], ['6/23/2016 22:20', 2], ['6/17/2016 0:01', 3], ['9/30/2015 4:12', 8]] 

{'11': 660, '19': 980, '22': 758, '00': 611, '04': 454, '09': 534, '16': 1101, '18': 1084, '10': 591, '12': 789, '20': 911, '03': 407, '17': 1169, '14': 958, '13': 918, '01': 500, '23': 674, '08': 496, '02': 441, '21': 874, '15': 1040, '06': 408, '05': 388, '07': 448} 

{'11': 37995, '19': 58811, '22': 38079, '00': 35718, '04': 22549, '09': 28802, '16': 59655, '18': 58459, '10': 35746, '12': 45287, '20': 41218, '03': 23167, '17': 67777, '14': 59191, '13': 57398, '01': 25303, '23': 35068, '08': 26830, '02': 25786, '21': 43149, '15': 62964, '06': 18864, '05': 19387, '07': 25461} 

[['11', 57.57], ['19', 60.01], ['22', 50.24], ['00', 58.46], ['04', 49.67], ['09', 53.94], ['16', 54.18], ['18', 53.93], ['10', 60.48], ['12', 57.4], 

In our comprehensive analysis of Hacker News posts, we also examined posts that were neither labeled as "Ask HN" nor "Show HN." Here are our key findings:

- Our analysis looked at general posts on Hacker News, excluding "Ask HN" and "Show HN" posts, and found the following:

- General Post Statistics:
    - There are 17,194 of these general posts.
    - They received a total of 462,055 comments, averaging about 27 comments per post.
    - They also received a total of 952,664 points, averaging about 55 points per post.
- Best times for comments:
    - 2 PM: 32 comments on average
    - 1 PM: 31 comments on average
    - 12 PM: 30 comments on average
    - 11 AM: 30 comments on average
    - 3 PM: 30 comments on average
- Best times for points:
    - 1 PM: 63 points on average
    - 2 PM: 62 points on average
    - 3 PM: 61 points on average
    - 10 AM: 60 points on average
    - 7 PM: 60 points on average

This analysis indicates that general posts on Hacker News receive the most engagement in terms of both comments and points when posted around midday and early afternoon.

### Comparison of "Ask HN", "Show HN", and Other Posts

In our analysis of Hacker News posts, we compared three categories: "Ask HN," "Show HN," and other posts. Here are our summarized findings:

Comments:
- "Ask HN" posts generate the most comments in the afternoon, peaking at 3 PM.
- "Show HN" posts receive fewer comments compared to "Ask HN" and other posts, with peak engagement in the morning and afternoon.
- Other posts generally receive the highest number of comments, especially around midday (12 PM to 3 PM).

Points:
- "Ask HN" posts receive the most points in the afternoon, with 3 PM being the optimal time.
- "Show HN" posts perform best late at night or around noon, receiving the highest points at 11 PM.
- Other posts get the most points overall, with peak times from late morning to early afternoon (10 AM to 3 PM).

Overall, our analysis indicates that while "Ask HN" posts are effective for generating comments, "Show HN" posts excel in gaining points, particularly when posted at specific times. Other posts outperform both "Ask HN" and "Show HN" posts in both comments and points, especially when posted around midday.